python/lsst/daf/butler/registry/datasets/byDimensions/_storage.py

```diff
@@ -430,3 +402,200 @@ def getDataId(self, id: int) -> DataCoordinate:
+            assert datasetIds is not None
+            # Update the summary tables for this collection in case this is the
+            # first time this dataset type or these governor values will be
+            # inserted there.
+            self._summaries.update(run, self.datasetType, self._dataset_type_id, governorValues)
+            # Combine the generated dataset_id values and data ID fields to
+            # form rows to be inserted into the tags table.
+            protoTagsRow = {
+                "dataset_type_id": self._dataset_type_id,
+                self._collections.getCollectionForeignKeyName(): run.key,
+            }
+            tagsRows = [
+                dict(protoTagsRow, dataset_id=dataset_id, **dataId.byName())
+                for dataId, dataset_id in zip(dataIdList, datasetIds)
+            ]
+            # Insert those rows into the tags table.  This is where we'll
+            # get any unique constraint violations.
+            self._db.insert(self._tags, *tagsRows)
+        for dataId, datasetId in zip(dataIdList, datasetIds):
+            yield DatasetRef(
+                datasetType=self.datasetType,
+                dataId=dataId,
+                id=datasetId,
+                run=run.name,
+            )
+
+    def import_(self, run: RunRecord, datasets: Iterable[DatasetRef],
+                idGenerationMode: DatasetIdGenEnum = DatasetIdGenEnum.UNIQUE) -> Iterator[DatasetRef]:
+        # Docstring inherited from DatasetRecordStorage.
+
+        # We do not reuse integer IDs from other repository as it is going
+        # to clash with auto-incremented IDs from this and other repos.
+        return self.insert(
+            run,
+            (dataset.dataId for dataset in datasets),
+            idGenerationMode
+        )
+
+
+class ByDimensionsDatasetRecordStorageUUID(ByDimensionsDatasetRecordStorage):
+    """Implementation of ByDimensionsDatasetRecordStorage which uses UUID for
+    dataset IDs.
+    """
+
+    NS_UUID = uuid.UUID('840b31d9-05cd-5161-b2c8-00d32b280d0f')
+    """Namespace UUID used for UUID5 generation. Do not change. This was
+    produced by `uuid.uuid5(uuid.NAMESPACE_DNS, "lsst.org")`.
+    """
+
+    def insert(self, run: RunRecord, dataIds: Iterable[DataCoordinate],
+               idMode: DatasetIdGenEnum = DatasetIdGenEnum.UNIQUE) -> Iterator[DatasetRef]:
+        # Docstring inherited from DatasetRecordStorage.
+
+        # Iterate over data IDs, transforming a possibly-single-pass iterable
+        # into a list, and remembering any governor dimension values we see.
+        dataIdList = []
+        rows = []
+        for dataId in dataIds:
+            dataIdList.append(dataId)
+            rows.append({
+                "id": self._makeDatasetId(run, dataId, idMode),
+                "dataset_type_id": self._dataset_type_id,
+                self._runKeyColumn: run.key,
+            })
+
+        yield from self._insert(run, dataIdList, rows, self._db.insert)
```

Comment:
If you want existing datasets with the same deterministic ID to be ignored rather than cause conflict->rollback (as the docs say), I think you want to use `self.db.ensure` here.   Same in `import_` below.  And ideally we'd then inspect the number of rows inserted (which `ensure` will return) to make sure it's the same as what we passed in, unless `idMode` is a deterministic one - that would catch collisions, which I guess we shouldn't worry about (but I can't help myself).

But there's still a problem with that - there's also a `db.insert` into the tags table would fail due to the unique constraint on dataset type + data ID + collection.  If we use `db.ensure` there, too, _this_ problem is solved, but then we can't tell the difference between inserting the exact same dataset twice and trying to insert a genuinely conflicting dataset with a different `dataset_id` but the same dataset type + data ID + collection.

So I think the easiest way out by far is to change all of the new docs for now that advertise that inserting the same dataset will do nothing.  I think the alternative is to both add a new constraint to the `dataset_tags_*` tables and upgrade `db.ensure` to pay attention to which constraint failed (something I think PostgreSQL can already do, but SQLite can only do in the very latest version that we just blocked upgrading to because of performance regressions).

Reply:
This is one unanswered question that I had on JIRA. It is not quite clear to me whether `insert` should skip existing dataset or raise. After some thinking I decided that raising is probably the right thing to do, my idea is that ingesting the same (raw) dataset into the same repo should not happen in general, and we should only need to care about identical datasets on import. If this is not true and we actually want `insert` to behave exactly as `import` then `ensure` is the right thing to do. There is indeed an issue of which constraint is violated with `ensure` - PK/ID or uniqueness, this I think matters for detecting genuine ID collisions, I'm not sure how to handle this best.

Reply:
:+1: for raising.  Please just check that this is consistent with the documentation.

Reply:
`insertDataset()` already documents that exception is raised when attempting to insert dataset with matching dataId/dataset/run.